apps/webapp/src/fetchers/staking.ts

```diff
@@ -1,6 +1,84 @@
+  getDisplayDenomFromView,
+  getIdentityKeyFromValidatorInfo,
+  getValidatorInfo,
+} from '@penumbra-zone/types';
+import { DelegationCaptureGroups, assetPatterns } from '@penumbra-zone/constants';
+import { Any } from '@bufbuild/protobuf';
+import { ValidatorInfo } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/stake/v1/stake_pb';
+
+const isDelegationBalance = (assetBalance: AssetBalance, identityKey: IdentityKey) => {
+  const match = assetPatterns.delegationToken.exec(getDisplayDenomFromView(assetBalance.value));
+  if (!match) return false;
+
+  const matchGroups = match.groups as unknown as DelegationCaptureGroups;
+
+  return bech32IdentityKey(identityKey) === matchGroups.bech32IdentityKey;
+};
 
 /**
+ * Given an `AddressIndex`, yields `ValueView`s of the given address's balance
+ * of delegation tokens. Each `ValueView` has an `extendedMetadata` property
+ * containing the `ValidatorInfo` for the validator the delegation token is
+ * staked in.
+ *
+ * Note that one `ValueView` will be returned for each validator, even if the
+ * given address holds no tokens for that validator. (When there are no tokens
+ * for a given validators, the `ValueView` will have an amount of zero.) This
+ * ensures that you can display all validators by iterating over the response
+ * from this method, rather than having to call one method to get validators the
+ * user has stake in, and another for validators the user doesn't have stake in.
+ *
+ * @todo: Make this an RPC method, rather than doing it in the webapp.
  * @todo: Make `showInactive` configurable via UI filters.
  */
-export const getValidatorInfos = () => stakingClient.validatorInfo({ showInactive: false });
+export const getDelegationsForAccount = async function* (addressIndex: AddressIndex) {
+  const assetBalances = await getAssetBalances({ accountFilter: addressIndex });
+  const validatorInfoResponses = stakingClient.validatorInfo({ showInactive: false });
+
+  for await (const validatorInfoResponse of validatorInfoResponses) {
+    const extendedMetadata = new Any({
+      typeUrl: ValidatorInfo.typeName,
+      value: validatorInfoResponse.validatorInfo?.toBinary(),
+    });
+
+    const identityKey = getValidatorInfo.pipe(getIdentityKeyFromValidatorInfo)(
+      validatorInfoResponse,
+    );
+    const delegation = assetBalances.find(assetBalance =>
+      isDelegationBalance(assetBalance, identityKey),
+    );
+
+    if (delegation) {
+      const withValidatorInfo = delegation.value.clone();
+
+      if (withValidatorInfo.valueView.case !== 'knownAssetId')
+        throw new Error(`Unexpected ValueView case: ${withValidatorInfo.valueView.case}`);
+
+      withValidatorInfo.valueView.value.extendedMetadata = extendedMetadata;
+
+      yield withValidatorInfo;
+    } else {
+      yield new ValueView({
+        valueView: {
+          case: 'knownAssetId',
+          value: {
+            amount: {
+              hi: 0n,
+              lo: 0n,
+            },
```

Comment:
Not sure what to populate `metadata` with here, so I left it `undefined`. Any suggestions? (I'd prefer to punt that to a future PR, though.)

Reply:
ideally it should be the token metadata for the delegation, from the extension's token registry -- and we'd really like that info to take priority over the validator info if possible, because it's data we check in to the extension's source code rather than just something someone posted to the chain

Reply:
You can do

```
const { denomMetadata } = await viewClient.assetMetadataById({ assetId });
```

but it's not super clear to me where you'd get this assetId if it's in the `!delegation` part of the clause.

Reply:
@grod220 Yeah, exactly — I'm not sure how to look that up since we don't actually have any tokens. I'll punt this down to a later PR and make a ticket for it.

Reply:
[Ticket](https://github.com/penumbra-zone/web/issues/617)